### ISSUE TITLE
User lookup failures

### ISSUE THREAD

**Issue by manojampalam** [PowerShell/Win32-OpenSSH#553]
This is to keep track of bunch of related issues reported around failures to resolve provided user name. Issue is within this logic in pwd.c

if ((status = NetUserGetInfo(udom_utf16, uname_utf16, 23, &user_info)) != NERR_Success) {
			debug("NetUserGetInfo() failed with error: %d for user: %ls and domain: %ls \n", status, uname_utf16, udom_utf16);

			if ((dsStatus = DsGetDcNameW(NULL, udom_utf16, NULL, NULL, DS_DIRECTORY_SERVICE_PREFERRED, &pdc)) != ERROR_SUCCESS) {
				error("DsGetDcNameW() failed with error: %d \n", dsStatus);
				errno = ENOENT;
				goto done;
			}

			if ((status = NetUserGetInfo(pdc->DomainControllerName, uname_utf16, 23, &user_info)) != NERR_Success) {
				debug("NetUserGetInfo() with domainController: %ls failed with error: %d \n", pdc->DomainControllerName, status);
				errno = ENOENT;
				goto done;
			}

**Comment by manojampalam** [comment]
Related issues:
https://github.com/PowerShell/Win32-OpenSSH/issues/378
https://github.com/PowerShell/Win32-OpenSSH/issues/543
https://github.com/PowerShell/Win32-OpenSSH/issues/543

**Comment by jasongin** [comment]
I hit this problem on a Windows 10 system joined to Azure AD ("work or school account").

    NetUserGetInfo() failed with error: 5

After I disconnected the system from Azure AD and joined a regular local network domain, then OpenSSH authentication succeeded.

**Comment by hume-github** [comment]
I will add that I'm seeing the above error ("NetUserGetInfo() failed with error: 5") on a Windows 10 PC joined to a Win2008R2 domain, but *only* when sshd is running as a service.  If I start sshd from an Administrator powershell, I can SSH in with my domain account just fine.

A wireshark shows the domain controller returning a "OpenUser response, STATUS_ACCESS_DENIED".

**Comment by archiker** [comment]
Any luck to get this fixed/workaround?

I am using Win2k12 server within domain with sshd service, and SFTP login with password auth from Win7 WinSCP client. Same error as above: https://github.com/PowerShell/Win32-OpenSSH/issues/1142

Thanks.

**Comment by manojampalam** [comment]
We have a fix in pipeline from @nomorefood.  This should be fixed in next release.

**Comment by manojampalam** [comment]
@nomorefood with the new group fetch logic in getusergroups(), local built in groups are being fetched in the form builtin\group (ex. builtin\administrators). 

Did you have any specific reason to keep the builtin prefix ? If not, can we remove it, to keep it consistent with "user" rules (local user have no domain prefix). This will also align it with how local groups are commonly used in various utilities
Ex. net localgroup administrators

**Comment by NoMoreFood** [comment]
@manojampalam I had thought about trimming it off but I didn't know if 'BUILTIN' is localization-sensitive?  I'll put in a PR to remove it if you happen to know for sure (or have a non-English platform to confirm on).

**Comment by manojampalam** [comment]
@NoMoreFood let me follow up. 
I'm also facing one other issue. With the following entry:
Match Group administrator
I'm getting up a 14 second delay on my enterprise box (my account has ~250 group memberships)

I'm planning on doing the following optimization:
If the group name does not have wild cards, I'll resolve the name to SID and simply do a membership check against the token. This will avoid enumerating all the group names and save significant time. Let me do the change and I'll also address dropping BUILTIN as appropriate. Will update you when I have the changes ready

**Comment by NoMoreFood** [comment]
@manojampalam Sounds good.  Some of that delay is probably due to the de-optimization I had to do for OneCore compatibility.  Since the APIs that allow simultaneous, multiple account resolution are not exposed through OneCore, I had to change it to a one-by-one method.  The annoying thing is that at least on the Windows desktop/server platforms, the OneCore-endorsed APIs end up just calling the ones I was previously using and ultimately just passing in a SID to resolve one at a time.  Maybe it's different on other platforms....
